namenode/namenode.go

```diff
@@ -44,27 +55,280 @@ func (server *NameNode) Create(path string, success *bool) error {
+		for endpoint, dataNode := range server.datanodes {
+			if !dataNode.heartbeatReceived {
+				// The datanode hasn't sent a heartbeat
+				server.datanodes[endpoint] = DataNodeInfo{
+					heartbeatReceived: false,
+					isAlive:           false,
+				}
+			} else {
+				server.datanodes[endpoint] = DataNodeInfo{
+					heartbeatReceived: false,
+					isAlive:           true,
+				}
+			}
+		}
+		server.datanodeRWLock.Unlock()
+	}
+}
+
+// ReadRequest handles the read request from the client,
+// and returns (Block Name, DataNode Endpoint) pair for each block
+// TODO: only return DataNode Endpoint ?
+func (server *NameNode) ReadRequest(args *common.ReadFileRequest, reply *common.ReadFileResponse) error {
+	slog.Info("ReadRequest", "file", args.FileName)
+
+	server.inodeRWLock.RLock()
+	inode, exists := server.inodes[args.FileName]
+	server.inodeRWLock.RUnlock()
+
+	if !exists {
+		reply.Succeeded = false
+		slog.Error("file not found", "file", args.FileName)
+		return errors.New("file not found")
+	}
+
+	// TODO: if DataNode status changed during request, the client should resend the request
+	dataNodesStatus := server.getDataNodesStatus()
+
+	inode.rwlock.RLock()
+	defer inode.rwlock.RUnlock()
+
+	var blockInfoList []common.BlockInfo
+	for i := args.BlockRange[0]; i < args.BlockRange[1]; i++ {
+		storageInfo, exist := inode.storageInfo[i]
+		if !exist {
+			reply.Succeeded = false
+			slog.Error("block index does not exist", "block index", i)
+			return errors.New("block index does not exist")
+		}
+
+		storageNodes := storageInfo.DataNodes
+		// Filter out the alive storage nodes
+		var hasDeadNode bool = false
+		var aliveStorageNodes []string
+		for _, storageNode := range storageNodes {
+			if dataNodesStatus[storageNode] {
+				aliveStorageNodes = append(aliveStorageNodes, storageNode)
+			} else {
+				hasDeadNode = true
+			}
+		}
+		// Update the storage info if needed
+		if hasDeadNode {
+			inode.rwlock.RUnlock()
+			inode.rwlock.Lock()
+			storageInfo.DataNodes = aliveStorageNodes
+			inode.storageInfo[i] = storageInfo
+			inode.rwlock.Unlock()
+			inode.rwlock.RLock()
+		}
+		// // Randomly select an alive storage node
+		// if len(aliveStorageNodes) == 0 {
+		// 	reply.Succeeded = false
+		// 	slog.Error("no alive DataNode")
+		// 	return errors.New("no alive DataNode")
+		// }
+		// dataNode := aliveStorageNodes[rand.Intn(len(aliveStorageNodes))]
+
+		blockname := common.ConstructBlockName(args.FileName, i, inode.storageInfo[i].LatestVersion)
+		// TODO: should we not disclose version number to client?
+		blockInfoList = append(blockInfoList, common.BlockInfo{
+			BlockName:         blockname,
+			DataNodeEndpoints: aliveStorageNodes,
+		})
+	}
+
+	reply.Succeeded = true
+	reply.BlockInfoList = blockInfoList
+
+	slog.Debug("ReadRequest success", "file", args.FileName, "blockInfoList", blockInfoList)
+
+	return nil
+}
+
+// Returns the status of all datanodes
+func (server *NameNode) getDataNodesStatus() map[string]bool {
```

Comment:
This method name is obscure

Reply:
If you only want to know whether a datanode is alive, then why not name is `getDataNodeLiveness`